File: minDepth.py
class Solution:
# Approach 1- DFS (Recursive) [using implicit call-stack]
    def minDepth(self, root): # TC O(n) // SC O(n)
        if not root: 
            return 0 
        
        children = [root.left, root.right]
        # if we're at leaf node
        if not any(children):
            return 1
        
        min_depth = float('inf')
        for c in children:
            if c:
                min_depth = min(self.minDepth(c), min_depth)
        return min_depth + 1 

# Approach 2- BFS [using queue]
    def minDepth1(self, root): # TC O(n) // SC O(n)
        if not root: return 0
        q=deque()
        q.append(root)
        depth=0
        while q:
            #currNode=q.popleft()
            depth+=1
            for node in range(len(q)):
                currNode=q.popleft()
                if not currNode.left and not currNode.right:
                    return depth
                if currNode.left: q.append(currNode.left)
                if currNode.right: q.append(currNode.right)
            
        
        return depth

s=Solution()
print(s.min_depth(root))

File: minDepth.java
import java.util.*;
// Approach 1- DFS (Recursive) [using implicit call-stack]
class Solution{ // TC O(n) SC O(n)
  public int minDepth(Node root) {
    if (root == null) {
      return 0;
    }

    if ((root.left == null) && (root.right == null)) {
      return 1;
    }

    int min_depth = Integer.MAX_VALUE;
    if (root.left != null) {
      min_depth = Math.min(minDepth(root.left), min_depth);
    }
    if (root.right != null) {
      min_depth = Math.min(minDepth(root.right), min_depth);
    }

    return min_depth + 1;
  }
}

// Approach 2- BFS [using queue]
class Solution1{ // TC O(n) SC O(n)
    public int minDepth(Node root) {
        if(root==null) return 0;
        Queue<Node> queue=new LinkedList<>();
        queue.offer(root);
        int depth=0;
        while(!queue.isEmpty()){
            int size=queue.size();
            depth++;
            for(int i=0; i<size; i++){
                Node currNode= queue.poll();
                if (currNode.left==null && currNode.right==null){
                    return depth;
                }
                if (currNode.left!=null) queue.offer(currNode.left);
                if (currNode.right!=null) queue.offer(currNode.right);
            }
        }
        return depth;
    }
}

class minDepth{
  public static void main(String[] args) {
    BinaryTree bt=new BinaryTree();
    bt.root=new Node(1);
    bt.root.left=new Node(2);
    bt.root.right=new Node(3);
    bt.root.right.right=new Node(4);
    Solution s=new Solution();
    System.out.println(s.minDepth(bt.root));
  }
}
